### ISSUE TITLE
Too many database connections

### ISSUE THREAD

**Issue by bnmnetp** [RunestoneInteractive/rs#364]

**Comment by bnmnetp** [comment]
Each worker in the runestone swarm creates 65 database connections.  This is too many and leads to postgresql issues.  

The problem is that we create the session in the `async_session.py` file.  which is then imported nearly anywhere.


 --  try this [python - How to do persistent database connection in FastAPI? - Stack Overflow](https://stackoverflow.com/questions/63270196/how-to-do-persistent-database-connection-in-fastapi) and also [How to maintain global pool of db connection and use it in each and every request? · Issue #1800 · tiangolo/fastapi · GitHub](https://github.com/tiangolo/fastapi/issues/1800) -- essentially this moves the session object to request.state -- which seems good, but would require adding a parameter to everything in crud.py -- or perhaps revamping crud.py to be a class and storing async_session as an instance variable initialized from request.state

**Comment by bjones1** [comment]
Sorry for the slow response! I'm heavily in favor of option 2. That makes so much sense and keeps the code a lot DRYer.

**Comment by bnmnetp** [comment]
@bjones1 -- I have a prototype of option 2 on a branch.  But there are some challenges.

1. There are definitely places where we do not have a request but we still want to use crud functions.
2. I think there are still places in the code where we will leak sessions unless we are **very** careful about imports.

A different solution came to me from mlhess at Umich.  He uses `pgbouncer` with his kubernetes setup.  pgbouncer is a dedicated connection pooler for postgresql. (pgbouncer.org)  We set up our database connection to connect to a local copy of `pgbouncer` and it takes care of managing the connection pool and queuing requests.  This would allow us to set a carefully chosen pool size for each worker.  This feels really clean to me.

Any thoughs

**Comment by bjones1** [comment]
I like this option, but does pgbouncer do what we want? If we use the session pooling mode (the default/most compatible), then it seems like it still allocates one server connection per client connection. I'm assuming we connect to the db on startup and don't terminate until shutdown.

**Comment by bnmnetp** [comment]
Yes.  Each worker would run a pgbouncer.  Each pgbouncer would have 20 (configurable) connections to the database.  The servers on the worker connect to pgbouncer.  If there are more than 20 then it queues the requests before passing them on to the postgresql server.

**Comment by bjones1** [comment]
Sounds great to me! Let me know what happens.

**Comment by bnmnetp** [comment]
Closing as resolved. We went with the pgbouncer approach rather than threading a session object through `crud.py`: `docker-compose.yml` now runs a pgbouncer service, and the book, assignment, and admin servers connect through it under the `production` profile. Each worker's connections are pooled and queued instead of multiplying out to the database.

If connection exhaustion turns up again in production, a fresh issue with the actual numbers will be more useful than reopening this.
